llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,320 @@
+  /// accessed, i.e. the maximum value which was used as argument of the
+  /// resize method.
+  [[nodiscard]] size_t size() const { return Size; }
+
+  /// Resize the vector. Notice that the constructor of the elements will not
+  /// be invoked until an element of a given page is accessed, at which point
+  /// all the elements of the page will be constructed.
+  ///
+  /// If the new size is smaller than the current size, the elements of the
+  /// pages that are not needed anymore will be destroyed, however, elements of
+  /// the last page will not be destroyed.
+  ///
+  /// For these reason the usage of this vector is discouraged if you rely
+  /// on the construction / destructor of the elements to be invoked.
+  void resize(size_t NewSize) {
+    if (NewSize == 0) {
+      clear();
+      return;
+    }
+    // Handle shrink case: destroy the elements in the pages that are not
+    // needed anymore and deallocate the pages.
+    //
+    // On the other hand, we do not destroy the extra elements in the last page,
+    // because we might need them later and the logic is simpler if we do not
+    // destroy them. This means that elements are only destroyed only when the
+    // page they belong to is destroyed. This is similar to what happens on
+    // access of the elements of a page, where all the elements of the page are
+    // constructed not only the one effectively neeeded.
+    if (NewSize < Size) {
+      size_t NewLastPage = (NewSize - 1) / PageSize;
+      for (size_t I = NewLastPage + 1, N = PageToDataPtrs.size(); I < N; ++I) {
+        T *PagePtr = PageToDataPtrs[I];
+        if (PagePtr == InvalidPage)
+          continue;
+        T *Page = PagePtr;
+        // We need to invoke the destructor on all the elements of the page.
+        std::destroy_n(Page, PageSize);
+        Allocator.getPointer()->Deallocate(Page);
+        // We mark the page invalid, to avoid double deletion.
+        PageToDataPtrs[I] = InvalidPage;
+      }
+      PageToDataPtrs.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages. Notice that we do not
+    // need to default construct any new element, because that was already done
+    // when the page was allocated.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PageSize.
+    // In that case we need one more page.
+    size_t Pages = Size / PageSize;
+    size_t Remainder = Size % PageSize;
+    if (Remainder != 0)
+      Pages += 1;
+    assert(Pages > PageToDataPtrs.size());
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    PageToDataPtrs.resize(Pages, InvalidPage);
```

Comment:
```suggestion
    // We use InvalidPage to indicate that a page has not been allocated yet.
    // We use InvalidPage instead of a separate bool to avoid wasting space.
    PageToDataPtrs.resize(NewLastPage + 1, InvalidPage);
```

- We don't need this complicated page calculation, it's doing the same thing as the simpler `NewLastPage` calculation.
- `0` (now `nullptr`) isn't a valid page id any more, and is actually the same as `InvalidPage`.

Reply:
As I wrote above, I would prefer keep the explicit InvalidPage for now, since I might have a future development where it goes back to be -1.

Reply:
That is not what I'm suggesting that you change here.

Reply:
Ok, but then I think I already did the change you suggested:

https://github.com/llvm/llvm-project/pull/66430/files#diff-073e73a590a55c746db8abc3ace75b706d2033ff6134ba81867aeeb27f78bea3R165-R171